libs/chonky/src/embedding/multi_modal_embedding.rs

```diff
@@ -0,0 +1,338 @@
+        }
+        embeddings.push(page_embeddings);
+    }
+    Ok(embeddings)
+}
+
+/// Given the text on each page of the pdf, embeds each pages text seperately into a vector
+///
+/// # Errors
+///
+/// [`ChonkyError::VertexAPI`] when there are HTTP request errors
+pub async fn embed_text(
+    pdf_text_extract: Vec<Vec<String>>,
+    project_id: &str,
+) -> Result<Vec<Vec<f64>>, Report<ChonkyError>> {
+    let mut embeddings = Vec::new();
+    for page_text in pdf_text_extract {
+        embeddings.push(make_multimodal_api_request(project_id, None, Some(page_text)).await?);
+    }
+    Ok(embeddings)
+}
+
+/// A function that performs authentication with Google Vertex API and performs
+/// a curl request to obtain multimodal embeddings given an image path
+///
+/// # Errors
+///
+/// [`ChonkyError::VertexAPI`] when there are HTTP request errors
+/// [`ChonkyError::ImageError`] when there are errors converting to base64 encoding
+pub async fn make_multimodal_api_request(
+    project_id: &str,
+    image_data: Option<Vec<u8>>,
+    text_data: Option<Vec<String>>,
+) -> Result<Vec<f64>, Report<ChonkyError>> {
+    let url = format!(
+        "https://us-central1-aiplatform.googleapis.com/v1/projects/{project_id}/locations/us-central1/publishers/google/models/multimodalembedding@001:predict"
+    );
+
+    let access_token = get_vertex_access_token()?; // Assuming this function is synchronous
+
+    // Create the reqwest async client
+    let client = Client::new();
+
+    // Prepare headers
+    let mut headers = HeaderMap::new();
+    headers.insert(
+        AUTHORIZATION,
+        HeaderValue::from_str(&format!("Bearer {access_token}"))
+            .change_context(ChonkyError::VertexAPI)?,
+    );
+    headers.insert(
+        CONTENT_TYPE,
+        HeaderValue::from_static("application/json; charset=utf-8"),
+    );
+
+    // Prepare payload
+    let mut payload = String::new();
+    if let Some(image_payload) = image_data {
+        payload = base64_json(image_payload)?; // Assuming this function prepares the JSON
+    } else if let Some(text_payload) = text_data {
+        payload = text_json(&text_payload)?; // Assuming this function prepares the JSON
+    }
+
+    // Make the POST request
+    let response = client
+        .post(&url)
+        .headers(headers)
+        .body(payload)
+        .send()
+        .await
+        .change_context(ChonkyError::VertexAPI)?;
+
+    // Check the response status
+    if !response.status().is_success() {
+        return Err(Report::new(ChonkyError::VertexAPI));
+    }
+
+    // Read and process the response
+    let response_text = response
+        .text()
+        .await
+        .change_context(ChonkyError::VertexAPI)?;
+
+    extract_embedding(&response_text)
+}
+
+// Parses the response to extract the image embedding vector
+fn extract_embedding(response_text: &str) -> Result<Vec<f64>, Report<ChonkyError>> {
```

Comment:
Was implementing this but the `ImageEmbedding` struct requires the image in the struct, which is not accessible when just requesting the embedding, since the calling function should have knowledge on the type of vector it is receiving, it may be easier to not implement this (to avoid having to handling `if let` branches  upstream where the compiler does not know what datatype we are expecting but we do)